### DISCUSSION TITLE
Initialize a class-based provider with parameters

### DISCUSSION THREAD

**Discussion by ToddManUtah** [rrousselGit/riverpod#2709] [Q&A] [answered]
I have a class-based provider that requires a parameter. My question is how to initialize it? I thought I would have to create the provider first and then use the overrides: feature but overridesWithValue doesn't seem to be defined. I'm sure I'm doing something wrong but I don't know what. Is there a way to initialize a class-based provider WITHOUT using the overrides feature?

Here's the code I'm using to generate the provider:

    import 'package:riverpod_annotation/riverpod_annotation.dart';
    import 'package:dio/dio.dart';
    import 'package:flutter/foundation.dart';
    import 'package:shopping_list/models/grocery_item.dart';
    
    part 'grocery_list_repository_provider.g.dart';
    
    @riverpod
    class GroceryListRepository extends _$GroceryListRepository {
      late Dio client;
      late String apiUrl; //  = 'https://grocery-list-xxxxxx.firebaseio.com';
    
      @override
      Future<List<GroceryItem>> build({required String apiUrl}) async {
        this.apiUrl = apiUrl;
        client = Dio();
        return await loadGroceryList();
      }
    
      Future<List<GroceryItem>> loadGroceryList() async {
        try {
          final response = await client.get('$apiUrl/grocery-list');
          if(response.statusCode != 200) {
            return [];
          }
          final List<dynamic> json = response.data;
          return json.map((dynamic item) => GroceryItem.fromJson(item)).toList();
        } catch (e) {
          if (kDebugMode) {
            print('Error fetching grocery items: $e');
          }
          return [];
        }
      }
    
      Future<bool> addGroceryItem(GroceryItem item) async {
        final response = await client.post('$apiUrl/grocery-list', data: item.toJson());
        return response.statusCode == 200;
      }
    
      Future<bool> removeGroceryItem(GroceryItem item) async {
        final response = await client.delete('$apiUrl/grocery-list/${item.id}');
        return response.statusCode == 200;
      }
    }

**Comment by rrousselGit** [comment, marked as answer]
Parameters are passed in the ref.watch expression:

```dart
ref.watch(groceryListRepositoryProvider(apiUrl: 'url'));
```
